etcdserver/snapshot_store.go

```diff
@@ -0,0 +1,138 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	dstorage "github.com/coreos/etcd/storage"
+)
+
+type snapshot struct {
+	r raftpb.Snapshot
+}
+
+func (s *snapshot) raft() raftpb.Snapshot { return s.r }
+
+func (s *snapshot) size() uint64 { return 0 }
+
+func (s *snapshot) writeTo(w io.Writer) (n int64, err error) { return 0, nil }
+
+func (s *snapshot) close() {}
+
+type snapshotStore struct {
+	// dir to save snapshot data
+	dir string
+	kv  dstorage.KV
+
+	// send empty to reqsnapc to notify the channel receiver to send back latest
+	// snapshot to snapc
+	reqsnapc chan struct{}
+	// a chan to receive the requested snapshot
+	// snapshotStore will receive from the chan immediately after it sends empty to reqsnapc
+	snapc chan raftpb.Snapshot
+
+	mu   sync.Mutex
+	snap *snapshot
+}
+
+func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
+	return &snapshotStore{
+		dir:      dir,
+		kv:       kv,
+		reqsnapc: make(chan struct{}),
+		snapc:    make(chan raftpb.Snapshot),
+	}
+}
+
+// getNew gets a new snapshot.
+// It returns ErrSnapshotTemporarilyUnavailable if it fails to return snapshot now.
+func (ss *snapshotStore) getNew() (*snapshot, error) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	if ss.snap != nil {
+		return nil, raft.ErrSnapshotTemporarilyUnavailable
+	}
+
+	// ask to generate v2 snapshot
+	ss.reqsnapc <- struct{}{}
+	// TODO: generate v3 snapshot at here
+	raftsnap := <-ss.snapc
+	ss.snap = &snapshot{
+		r: raftsnap,
+	}
+	return ss.snap, nil
+}
+
+// get gets the generated snapshot.
+func (ss *snapshotStore) get() *snapshot {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	return ss.snap
+}
+
+// del deletes the generated snapshot.
+func (ss *snapshotStore) del() {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	ss.snap.close()
+	ss.snap = nil
+}
+
+// save saves snapshot into disk.
+//
+// If snapshot has existed, it returns error and keeps the original snapshot
+// data. The function guarantees that if the call is aborted at any time,
+// it leaves either complete snapshot data or nothing.
+func (ss *snapshotStore) save(s *snapshot) error {
+	f, err := ioutil.TempFile(ss.dir, "tmp")
+	if err != nil {
+		return err
+	}
+	_, err = s.writeTo(f)
+	f.Close()
+	if err != nil {
+		return err
+	}
+	fn := path.Join(ss.dir, fmt.Sprintf("%016x.db", s.raft().Metadata.Index))
+	if fileutil.Exist(fn) {
+		return fmt.Errorf("snapshot to save has existed")
+	}
+	return os.Rename(f.Name(), fn)
+}
+
+// getSnapshotFilePath gets the file path of saved snapshot file.
```

Comment:
getSnapshotFilePath returns the file path for the snapshot with given index. If the snapshot does not exist...